framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -379,27 +379,74 @@ public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror
         }
 
         @Override
-        public boolean implementsWidening() {
-            return true;
+        public int numberOfIterationsBeforeWidening() {
+            return MAX_VALUES + 1;
         }
 
         @Override
-        public AnnotationMirror widenUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
-            AnnotationMirror lub = leastUpperBound(a1, a2);
+        public AnnotationMirror widenedUpperBound(
+                AnnotationMirror newQualifier, AnnotationMirror previousQualifier) {
+            AnnotationMirror lub = leastUpperBound(newQualifier, previousQualifier);
             if (AnnotationUtils.areSameByClass(lub, IntRange.class)) {
-                Range range = getRange(lub);
-                if (range.isWithin(Byte.MIN_VALUE, Byte.MAX_VALUE)) {
-                    return createIntRangeAnnotation(Range.BYTE_EVERYTHING);
-                } else if (range.isWithin(Short.MIN_VALUE, Short.MAX_VALUE)) {
-                    return createIntRangeAnnotation(Range.SHORT_EVERYTHING);
-                } else if (range.isWithin(Integer.MIN_VALUE, Integer.MAX_VALUE)) {
-                    return createIntRangeAnnotation(Range.INT_EVERYTHING);
+                Range lubRange = getRange(lub);
+                Range newRange = getRangeOrConvertIntVal(newQualifier);
+                Range oldRange = getRangeOrConvertIntVal(previousQualifier);
+                Range wubRange = widenRange(newRange, oldRange, lubRange);
+                return createIntRangeAnnotation(wubRange);
+            } else if (AnnotationUtils.areSameByClass(lub, ArrayLenRange.class)) {
+                Range lubRange = getRange(lub);
+                Range newRange = getRangeOrConvertArrayLen(newQualifier);
+                Range oldRange = getRangeOrConvertArrayLen(previousQualifier);
+                return createArrayLenRangeAnnotation(widenRange(newRange, oldRange, lubRange));
+            } else {
+                return lub;
+            }
+        }
+
+        private Range widenRange(Range newRange, Range oldRange, Range lubRange) {
+            if (newRange == null || oldRange == null) {
+                return lubRange;
+            }
+            // If the newRange closer to max than the oldRange, then move the from to the
```

Comment:
This sentence is rather hard to parse. I think an `is` is missing in the condition. Using mark-up for code would also disambiguate the use of `to` as field name or English word.